ACCESS_TOKEN, Widget

RonnyWinkler opened the pull request:
- Added possibility to use an ACCESS_TOKEN in env.json for local test
Changes in driver.compose.json (added loading pair screen to check ACCESS_TOKEN
Added token usage in api.ts

- Added Homeys widget for Home and Pulse device
Added API function in app.ts
Added realtime events in home and pulse device to send updates to the widget
Widget folder with widget definition and html elements

When you add a field ACCESS_TOKEN to env.json, this is used. If this is not set (official Tibber version, the pais screen should continue with the oAuth screen. Please check this before publishing the app (I tested without access token and client id and the oAuth screen was shown).

The widget can show energy data from Pulse device (assign the Pulse device in widget settings):
![grafik](https://github.com/user-attachments/assets/a0f436bb-8338-4087-8d97-3e03a765f18c)

The widget can show price information with or without chart (assign home device in widget settings):
![grafik](https://github.com/user-attachments/assets/e1a60a5e-73ad-4ba7-9e82-86adf091f46f)

The widget offers several options 
- show/hide chart
- zoom chart (yAxis)
- show simple line or colored chart
- color options for: color based on price level, color based on price range (1/5 of corrent range), vertical area color bases on price level.
![grafik](https://github.com/user-attachments/assets/6bcedb80-6559-42a0-81b8-e196a1c1123b)
![grafik](https://github.com/user-attachments/assets/c32f3c4c-6d2c-4fb4-80e9-98ff316e0cf3)
![grafik](https://github.com/user-attachments/assets/aca6d4b9-3263-4373-bce9-5d194a6681dd)

Referring to issue #102

RonnyWinkler committed:
ACCESS_TOKEN, Widget

- Added possibility to use an ACCESS_TOKEN in env.json for local test
- Added Homeys widget for Home and Pulse device

RonnyWinkler committed:
Widget

Fixed widget settings

RonnyWinkler committed:
chart layout

Improved chart layout

RonnyWinkler committed:
Merge pull request #1 from tibber/master

import commits from origin

RonnyWinkler commented:
Hi @toini 
What's the status of this PR? Can I help somehow to merge it? Thanks for taking a look at it :-)

toini commented:
Hi @RonnyWinkler! The PR is awaiting for the team owning this to review it. I've been working in another team for a few years now but I'll notify them about your question!

cscheiene commented:
> Hi @RonnyWinkler! The PR is awaiting for the team owning this to review it. I've been working in another team for a few years now but I'll notify them about your question!

Any update? Would be really useful to get the price widget into Homey

toini commented:
Sorry no, I did not get any response from the team.

jstaro commented:
Thanks for your hard work and being patient waiting for this belated response! The Homey app is sadly a bit neglected due to us being very busy in other areas. Apologies.

We really appreciate you taking the time to contribute.

We took a look at your PR. The dashboard widgets seem quite useful, especially the price chart.

We were curious if you've seen the recently announced [Energy Management features](https://homey.app/en-us/news/introducing-the-homey-energy-dongle/) that Athom is working on, and what your thoughts are in regards to the possible overlap here?

Judging by the screenshots, there are both

- An instantaneous power/accumulated energy chart. The update in v1.9.20 (in retrospect, we should have given this a 1.10.0 version number) graciously contributed by Athom themselves should already enable support for this

- A future prices chart. This feature is not out yet, it appears, and a bit unclear how it'll work, but we're hoping we can expose the Tibber prices there as a standard capability or label or something along those lines

In anticipation of these upcoming built-in features, we are somewhat apprehensive about pulling in this much code, especially HTML/JS/CSS with 3rd party charting dependencies. Furthermore we already are a bit resource constrained in maintaining the current app, and the custom nature of these widgets are visual enough to start to require actual UX and graphic design, which would necessitate pulling in more people and disciplines in the mix.

We were more hoping to be able to build upon the foundation Athom is laying here, with more standard capabilities that the entire Homey ecosystem can utilize and compose further.

Putting the Homey Energy Management overlap aside, we could definitely see the widgets be contained in a separate "companion" app without having to integrate it into the Tibber app proper (you probably also don't want to be reliant upon us having time to release a new version whenever you would like something changed).

If the Tibber app instead exposes the data in a general way for anyone to be able to make widgets like this, what additional data/capabilities do you see we would need to expose?

Apart from the widgets themselves, we do think the ability to use an access token for local development can be really handy, and hopefully make it easier for others to contribute, so we would definitely appreciate it if you would consider extracting that code, plus the debugger attaching code, out to a separate PR, and we would happily take a look at that separately.

The access token support for local development would also make it easier for you to maintain a fork with these widgets, if you so wish.

RonnyWinkler committed:
Removed widget, added app API

- Removed widget. It's provided as additional community app.
- Added app API instead to connect the widget app to Tibber app (read device list, register on realtime events)

RonnyWinkler committed:
Removed widget, added app API

RonnyWinkler committed:
Removed widget, added app API

RonnyWinkler commented:
- Removed widget. It's provided as additional community app.
- Added app API instead to connect the widget app to Tibber app (read device list, register on realtime events)

RonnyWinkler commented:
Hi @jstaro 
thanks for your answer and feedback. I fully understand your point not to include custom widget with the risk of further maintenance. It's better to make a community project for widgets.

I removed the widget from the app now.
Instead I added an app API to connect to Tibber app from another app.
In detail that means: 
- Tibber app offers two http endpoints to get a list of existing devices (Homey devices for Home and Pulse)
- Tibber app offers a http endpoint to trigger a realtime event to publish current Home data.
- The realtime events in Home/Pulse device is still present. This event send realtime data to event subscribers.
- The widget app has a permission to use Tibber app API (the user has to approve this when installing from app store).
- The widget app connect to the Tibber app API. When adding a widget the user selects his device (Home/Pulse). In this case, the devices are read from Tibber app.
- The widget app registers to realtime event. This way, the realtime data (Pule energy data, Home price data) is pushed to the widget app on every change.
- The widget app pushes the data to the widget that is refreshed in realtime.

This way, other app can also benefit using the realtime data if needed. And you don't have the ballast of widget and dependencies in your app.
I hope that's a good compromise :-)

RonnyWinkler committed:
Merge pull request #2 from tibber/main

Pull new Tibber version

jstaro commented on app.ts:
```diff
@@ -50,6 +64,42 @@ class TibberApp extends App {
     this.log('Tibber app is stopping');
   }
+  
+  // App API ============================================================================
+  async apiTriggerRealtimeData(){
+    // Trigger a realtime event for Home device to publish current data to listening apps
+    this.homey.drivers.getDriver('home').getDevices().forEach( (device ) => {
+      (device as HomeDevice).triggerRealtimeData();
+    });
+    return { success: true };
+  }
+
+  async apiGetHomeDevices(query: any){
+    // Return a list of Home devices matching the query string
+    let homes: { name: string; id: any; }[] = [];
+    let devices = this.homey.drivers.getDriver('home').getDevices();
+    devices.forEach(device => {
+        homes.push({
+          name: device.getName(),
+          id: device.getData().id
+        })
+    });
+    return homes.filter((item) => item.name.toLowerCase().includes(query.search.toLowerCase()));  
+  }
+
+  async apiGetPulseDevices(query: any){
+    // Return a list of Pulse devices matching the query string
+    let homes: { name: string; id: any; }[] = [];
+    let devices = this.homey.drivers.getDriver('pulse').getDevices();
+    devices.forEach(device => {
+        homes.push({
+          name: device.getName(),
+          id: device.getData().id
+        })
+    });
+    return homes.filter((item) => item.name.toLowerCase().includes(query.search.toLowerCase()));
+}
+
```
These methods could live under a property called `api`, so you could remove the `api` prefix from the methods,

```ts
api = {
  async triggerRealtimeData(){
    // Trigger a realtime event for Home device to publish current data to listening apps
    this.homey.drivers.getDriver('home').getDevices().forEach( (device ) => {
      (device as HomeDevice).triggerRealtimeData();
    });
    return { success: true };
  },
  // ...
}
```

jstaro reviewed (changes requested):
Looks overall great.
Some minor refactoring needed around the trigger method and the API parts.

This repo also use ESLint and Prettier, so if you could run a pass of that with
`npm run lint` that would be super.

It would be very nice if you could also document the `ACCESS_TOKEN` thing a tiny bit in the README. If you don't feel like it, we can do that.

RonnyWinkler committed:
Redesign based on PR feedback

RonnyWinkler commented on drivers/home/device.ts:
```diff
@@ -835,6 +848,15 @@ class HomeDevice extends Device {
+    } catch (err) {
+      console.error(err);
+    }
+  }
```
This function is redesigned. Instead of triggering a realtime event I added a simple "read data" endpoint. 
That's easier to understand and to use as a GET request.
The device function itself does not invoke the data update. It is only reading current device attributes and it returns the data JSON in similar syntax like it's done in realtime event.
This way, the widget can read current data on startup and later it gets realtime updates.

RonnyWinkler commented on app.ts:
```diff
@@ -50,6 +64,42 @@ class TibberApp extends App {
+        })
+    });
+    return homes.filter((item) => item.name.toLowerCase().includes(query.search.toLowerCase()));  
+  }
```
Updates are done :-)

RonnyWinkler commented on app.ts:
```diff
@@ -50,6 +64,42 @@ class TibberApp extends App {
+        })
+    });
+    return homes.filter((item) => item.name.toLowerCase().includes(query.search.toLowerCase()));
+}
```
Updates are done

RonnyWinkler commented on .homeycompose/app.json:
```diff
@@ -30,5 +30,19 @@
+    },
+    "getHomeDevices": {
+      "method": "get",
+      "path": "/home_devices/"
```
The API coding is removed from app.ts and included into api.ts. So API code is separated. I hope this is in your interest.

The API endpoints are basically http endpoints. 
The http GET on /home_devices is processed by the handler function getHomeDevices - so http method and endpoint are combined in function name.

https://apps.developer.homey.app/advanced/web-api

RonnyWinkler commented on .vscode/launch.json:
```diff
@@ -0,0 +1,35 @@
+{
+    "version": "0.2.0",
+    "configurations": [
+        {
+            "type": "node",
+            "request": "attach",
+            "restart": true,
+            "name": "Attach Tibber to HomeyPro 19",
+            "address": "192.168.1.13",
+            "port": 9291,
+            "localRoot": "${workspaceFolder}",
+            "remoteRoot": "/"
+        }, 
+        {
+            "type": "node",
+            "request": "attach",
+            "restart": true,
+            "name": "Attach Tibber to HomeyPro 23",
+            "address": "192.168.1.16",
+            "port": 9291,
+            "localRoot": "${workspaceFolder}",
+            "remoteRoot": "/app/"
+        },
+        {
+            "type": "node",
+            "request": "attach",
+            "restart": true,
+            "name": "Attach Tibber to HomeyPro 23 Docker",
+            "address": "192.168.1.42",
+            "port": 9229,
+            "localRoot": "${workspaceFolder}",
+            "remoteRoot": "/app/"
+        }
+    ]
+}
```
I just added the example to get it working. But with the added debug code in app.ts, you have to activate debugging when starting the app using `homey app run --remote` instead of `homey app install`.
So yes, it could also be removed. Let me know...

RonnyWinkler committed:
Redesign based on PR feedback

jstaro commented:
Superseded by #112

RonnyWinkler commented:
Hi, is there something that prevents releasing the test version as stable?
I still get comments because users are working ont the older stable version without widget support.
Thanks for checking.
Ronny
